File: backend/app/Models/BlogPost.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Support\Str;

class BlogPost extends Model
{
    use HasFactory;

    protected $fillable = [
        'title',
        'slug',
        'excerpt',
        'content',
        'category_id',
        'author_id',
        'featured_image',
        'featured',
        'status',
        'published_at',
        'read_time',
        'view_count',
        'meta_title',
        'meta_description',
        'tags',
    ];

    protected function casts(): array
    {
        return [
            'featured' => 'boolean',
            'published_at' => 'datetime',
            'read_time' => 'integer',
            'view_count' => 'integer',
            'tags' => 'array',
        ];
    }

    /**
     * Get the category that owns the blog post
     */
    public function category(): BelongsTo
    {
        return $this->belongsTo(BlogCategory::class);
    }

    /**
     * Get the author (user) who created the blog post
     */
    public function author(): BelongsTo
    {
        return $this->belongsTo(User::class, 'author_id');
    }

    /**
     * Scope to get only published posts
     */
    public function scopePublished($query)
    {
        return $query->where('status', 'published')
            ->where(function ($q) {
                $q->whereNull('published_at')
                    ->orWhere('published_at', '<=', now());
            })
            ->whereNotNull('category_id') // Ensure category exists
            ->whereNotNull('author_id'); // Ensure author exists
    }

    /**
     * Scope to get only featured posts
     */
    public function scopeFeatured($query)
    {
        return $query->where('featured', true);
    }

    /**
     * Scope to filter by category
     */
    public function scopeByCategory($query, $categoryId)
    {
        return $query->where('category_id', $categoryId);
    }

    /**
     * Scope to search posts
     */
    public function scopeSearch($query, $search)
    {
        return $query->where(function ($q) use ($search) {
            $q->where('title', 'like', "%{$search}%")
                ->orWhere('excerpt', 'like', "%{$search}%")
                ->orWhere('content', 'like', "%{$search}%");
        });
    }

    /**
     * Get the URL for this blog post
     */
    public function getUrlAttribute(): string
    {
        return url("/blog/{$this->id}");
    }

    /**
     * Calculate read time from content if not set
     */
    public function getCalculatedReadTimeAttribute(): int
    {
        if ($this->read_time) {
            return $this->read_time;
        }

        // Calculate based on content length (average reading speed: 200 words per minute)
        $wordCount = str_word_count(strip_tags($this->content ?? ''));
        return max(1, (int) ceil($wordCount / 200));
    }

    /**
     * Auto-generate slug from title if not provided
     */
    protected static function boot()
    {
        parent::boot();

        static::creating(function ($post) {
            if (empty($post->slug) && !empty($post->title)) {
                $post->slug = Str::slug($post->title);
            }
        });

        static::updating(function ($post) {
            if ($post->isDirty('title') && empty($post->slug)) {
                $post->slug = Str::slug($post->title);
            }
        });
    }

    /**
     * Increment view count
     */
    public function incrementViews(): void
    {
        $this->increment('view_count');
    }

    /**
     * Get the full URL for the featured image
     */
    public function getImageUrl(): ?string
    {
        if (!$this->featured_image) {
            return null;
        }

        // If it's already a full URL (external image), return as is
        if (str_starts_with($this->featured_image, 'http://') || str_starts_with($this->featured_image, 'https://')) {
            return $this->featured_image;
        }

        // Check if file exists in public disk (new uploads)
        $publicPath = storage_path('app/public/' . $this->featured_image);
        if (file_exists($publicPath)) {
            // File is in public storage, accessible via storage symlink
            $imagePath = 'storage/' . ltrim($this->featured_image, '/');
            return url($imagePath);
        }

        // Check if file exists in private disk (old files from before disk change)
        $privatePath = storage_path('app/private/' . $this->featured_image);
        if (file_exists($privatePath)) {
            // Serve private files via a route
            return route('blog.image', ['path' => base64_encode($this->featured_image)]);
        }

        // Fallback: assume file is in public storage (for cases where file exists but check failed)
        $imagePath = 'storage/' . ltrim($this->featured_image, '/');
        return url($imagePath);
    }
}

File: mobile_app/lib/features/blog/data/models/blog_post.dart
import 'blog_category.dart';
import 'blog_author.dart';
import '../../../../core/config/app_config.dart';

/// Blog Post Model
class BlogPost {
  final int id;
  final String title;
  final String slug;
  final String excerpt;
  final String content;
  final BlogCategory category;
  final BlogAuthor author;
  final String? image;
  final bool featured;
  final String status;
  final String? date;
  final String? publishedAt;
  final String? readTime;
  final int readTimeMinutes;
  final int viewCount;
  final String? metaTitle;
  final String? metaDescription;
  final List<String>? tags;
  final String? url;

  BlogPost({
    required this.id,
    required this.title,
    required this.slug,
    required this.excerpt,
    required this.content,
    required this.category,
    required this.author,
    this.image,
    required this.featured,
    required this.status,
    this.date,
    this.publishedAt,
    this.readTime,
    required this.readTimeMinutes,
    required this.viewCount,
    this.metaTitle,
    this.metaDescription,
    this.tags,
    this.url,
  });

  factory BlogPost.fromJson(Map<String, dynamic> json) {
    return BlogPost(
      id: json['id'] as int,
      title: json['title'] as String,
      slug: json['slug'] as String,
      excerpt: json['excerpt'] as String? ?? '',
      content: json['content'] as String? ?? '',
      category: BlogCategory.fromJson(json['category'] as Map<String, dynamic>),
      author: BlogAuthor.fromJson(json['author'] as Map<String, dynamic>),
      image: json['image'] as String?,
      featured: json['featured'] as bool? ?? false,
      status: json['status'] as String? ?? 'published',
      date: json['date'] as String?,
      publishedAt: json['published_at'] as String?,
      readTime: json['readTime'] as String?,
      readTimeMinutes: json['read_time'] as int? ?? 0,
      viewCount: json['view_count'] as int? ?? 0,
      metaTitle: json['meta_title'] as String?,
      metaDescription: json['meta_description'] as String?,
      tags: json['tags'] != null
          ? List<String>.from(json['tags'] as List)
          : null,
      url: json['url'] as String?,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'title': title,
      'slug': slug,
      'excerpt': excerpt,
      'content': content,
      'category': category.toJson(),
      'author': author.toJson(),
      'image': image,
      'featured': featured,
      'status': status,
      'date': date,
      'published_at': publishedAt,
      'readTime': readTime,
      'read_time': readTimeMinutes,
      'view_count': viewCount,
      'meta_title': metaTitle,
      'meta_description': metaDescription,
      'tags': tags,
      'url': url,
    };
  }

  /// Get full URL for image
  String? get imageUrl {
    if (image == null) return null;
    if (image!.startsWith('http://') || image!.startsWith('https://')) {
      return image;
    }
    final cleanPath = image!.startsWith('/') ? image!.substring(1) : image!;
    return '${AppConfig.apiBaseUrl}/$cleanPath';
  }

  /// Get full URL for author avatar
  String? get authorAvatarUrl {
    if (author.avatar == null) return null;
    if (author.avatar!.startsWith('http://') ||
        author.avatar!.startsWith('https://')) {
      return author.avatar;
    }
    final cleanPath = author.avatar!.startsWith('/')
        ? author.avatar!.substring(1)
        : author.avatar!;
    return '${AppConfig.apiBaseUrl}/$cleanPath';
  }
}
